Separate nullability flow analysis from DataFlowPass

cston opened the pull request:
Split `DataFlowPass` into two classes, with a common base class for assigning slots.

cston commented:
@dotnet/roslyn-compiler, @gafter, @AlekseyTs, @agocke please review.

gafter commented:
>         private readonly SmallDictionary<LocalFunctionSymbol, LocalFuncUsages> _localFuncVarUsages =

I suspect we will have to do something with local functions vs nullability. For now this is fine.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs:13 in a01088e. [](commit_id = a01088eb74ec5f8dfb45e160750a8be379399e2a, deletion_comment = False)

gafter commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.ObjectCreationPlaceholderLocal.cs:
```diff
@@ -8,7 +8,7 @@
 namespace Microsoft.CodeAnalysis.CSharp
 {
-    internal partial class DataFlowPass
+    internal partial class NullableWalker
```
>NullableWalker [](start = 27, length = 14)

Possibly want to change the file name too.

cston replied:
Renamed file.

---
In reply to: [148388390](https://github.com/dotnet/roslyn/pull/22932#discussion_r148388390) [](ancestors = 148388390)

cston commented:
@AlekseyTs the most recent commit should fix the behavior change.

AlekseyTs commented:
>         private MethodSymbol currentMethodOrLambda;

Should we start this name with ```_```?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:38 in 71eda1c. [](commit_id = 71eda1c5b5726cadc7608de35533d0150dad620e, deletion_comment = False)

AlekseyTs commented:
>             Debug.Assert(!IsConditionalState);

I think asserts like this should be moved to the base class so that more scenarios participate in verification.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:624 in 71eda1c. [](commit_id = 71eda1c5b5726cadc7608de35533d0150dad620e, deletion_comment = False)

AlekseyTs commented:
>         private void VisitStatementsWithLocalFunctions(BoundBlock block)

Can this code be shared?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:698 in 71eda1c. [](commit_id = 71eda1c5b5726cadc7608de35533d0150dad620e, deletion_comment = False)

AlekseyTs commented:
>         public override BoundNode VisitSwitchStatement(BoundSwitchStatement node)

It feels like this code can be moved to the base, and for other similar functions. 

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:763 in 71eda1c. [](commit_id = 71eda1c5b5726cadc7608de35533d0150dad620e, deletion_comment = False)

AlekseyTs commented:
>         private void DeclareVariable(LocalSymbol symbol)

This function should probably made abstract and moved to the base along with DeclareVariables.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:864 in 71eda1c. [](commit_id = 71eda1c5b5726cadc7608de35533d0150dad620e, deletion_comment = False)

AlekseyTs commented:
>             Debug.Assert(!IsConditionalState);

That assumes that derived classes call base, of course.

---
In reply to: [341874780](https://github.com/dotnet/roslyn/pull/22932#issuecomment-341874780) [](ancestors = 341874780)

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:624 in 71eda1c. [](commit_id = 71eda1c5b5726cadc7608de35533d0150dad620e, deletion_comment = False)

cston commented:
>         private void VisitStatementsWithLocalFunctions(BoundBlock block)

Done, thanks.

---
In reply to: [341874875](https://github.com/dotnet/roslyn/pull/22932#issuecomment-341874875) [](ancestors = 341874875)

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:698 in 71eda1c. [](commit_id = 71eda1c5b5726cadc7608de35533d0150dad620e, deletion_comment = False)

cston commented:
>         public override BoundNode VisitSwitchStatement(BoundSwitchStatement node)

Removed a couple of methods that simply delegated to the base class and added a `// PROTOTYPE` comment for other `Visit` methods.

---
In reply to: [341874977](https://github.com/dotnet/roslyn/pull/22932#issuecomment-341874977) [](ancestors = 341874977)

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:763 in 71eda1c. [](commit_id = 71eda1c5b5726cadc7608de35533d0150dad620e, deletion_comment = False)

cston commented:
>         private void DeclareVariable(LocalSymbol symbol)

Will make this `abstract` when moving `Visit` methods to base class.

---
In reply to: [341875070](https://github.com/dotnet/roslyn/pull/22932#issuecomment-341875070) [](ancestors = 341875070)

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:864 in 71eda1c. [](commit_id = 71eda1c5b5726cadc7608de35533d0150dad620e, deletion_comment = False)
